Sources/OfflineEntitlements/OfflineEntitlementsManager.swift

```diff
@@ -0,0 +1,106 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  OfflineEntitlementsManager.swift
+//
+//  Created by Nacho Soto on 3/22/23.
+
+import Foundation
+
+class OfflineEntitlementsManager {
+
+    private let deviceCache: DeviceCache
+    private let operationDispatcher: OperationDispatcher
+    private let api: OfflineEntitlementsAPI
+
+    init(deviceCache: DeviceCache,
+         operationDispatcher: OperationDispatcher,
+         api: OfflineEntitlementsAPI) {
+        self.deviceCache = deviceCache
+        self.operationDispatcher = operationDispatcher
+        self.api = api
+    }
+
+    func updateProductsEntitlementsCacheIfStale(
+        isAppBackgrounded: Bool,
+        completion: (@MainActor @Sendable (Result<(), Error>) -> Void)?
+    ) {
+        guard #available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *) else {
+            self.dispatchCompletionOnMainThreadIfPossible(completion, result: .failure(.notAvailable))
+            return
+        }
+
+        guard self.deviceCache.isProductEntitlementMappingCacheStale else {
+            self.dispatchCompletionOnMainThreadIfPossible(completion, result: .success(()))
+            return
+        }
+
+        Logger.debug(Strings.offlineEntitlements.product_entitlement_mapping_stale_updating)
+
+        self.api.getProductEntitlementMapping(withRandomDelay: isAppBackgrounded) { result in
+            switch result {
+            case let .success(response):
+                self.handleProductEntitlementMappingBackendResult(with: response)
+
+            case let .failure(error):
+                self.handleProductsEntitlementsUpdateError(error)
+            }
+
+            self.dispatchCompletionOnMainThreadIfPossible(
+                completion,
+                result: result
+                    .map { _ in () }
+                    .mapError(Error.backend)
+            )
+        }
+    }
+
+}
+
+extension OfflineEntitlementsManager {
+
+    enum Error: Swift.Error {
+
+        case backend(BackendError)
+        /// Offline entitlements require iOS 15+
+        case notAvailable
+
+    }
+
+}
+
+@available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+private extension OfflineEntitlementsManager {
+
+    func handleProductEntitlementMappingBackendResult(with response: ProductEntitlementMappingResponse) {
+        Logger.debug(Strings.offlineEntitlements.product_entitlement_mapping_updated_from_network)
```

Comment:
Small nitpick but I added this log after storing in the cache, to make sure it was logged only after the cache was updated. NABD though.

Reply:
Oh wait we have both already: `product_entitlement_mapping_stale_updating` and `product_entitlement_mapping_updated_from_network`.